### PR TITLE
openmpi: 4.1.6 -> 5.0.3

### PR DESCRIPTION
**Pull request by doronbehar**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by markuskowa** [comment]
Duplicate of https://github.com/NixOS/nixpkgs/pull/275796

**Comment by doronbehar** [comment]
The GitHub search has disappointed me...

https://github.com/NixOS/nixpkgs/pulls?q=openmpi%205.0

@markuskowa perhaps you'd be interested in incorporating the changes here into https://github.com/NixOS/nixpkgs/pull/275796 ? The declarative approach in the reformat commit in this PR is a significant improvement IMO.

**Comment by markuskowa** [comment]
> The GitHub search has disappointed me...
> 
> https://github.com/NixOS/nixpkgs/pulls?q=openmpi%205.0
> 

Yep, finding things in this big list of PRs becomes increasingly difficult.

> @markuskowa perhaps you'd be interested in incorporating the changes here into #275796 ? The declarative approach in the reformat commit in this PR is a significant improvement IMO.

Yes, although the update PR is already quite big (mostly due to the split output/.dev feature). I would like to get the update merged. This has been sitting around for way too long. I could easily integrate the `finalAttrs` and the `withFeatureAs/enableFeature`. More changes could go in a follow-up PR?

**Comment by doronbehar** [comment]
You did great job in https://github.com/NixOS/nixpkgs/pull/275796 but I managed to include the details I missed in this PR, and it came out nicer IMO - both in terms of the commit log, and the diff and the implementation details (also no merge conflict as well).

I also disentangled the plumed blas64 commit to a separate as that has no effect or isn't affected by the openmpi so it seems:

https://github.com/NixOS/nixpkgs/pull/327503

As for the mpi4py issue, I'm not sure, I'm still investigating whethere there are packages that actually, really, use this package - I suspect that many of the packages that depend upon it don't really use any functionality from there.

**Comment by markuskowa** [comment]
@GrahamcOfBorg build cp2k python3Packages.meep

**Comment by markuskowa** [comment]
@doronbehar thanks for integrating my changes in this PR. This looks pretty good now. Let me run nixpkgs-review on it to make sure we have everything covered.

On request: please put the reformatting in its own commit (i.e. put "and more" in a separate commit). That makes the changes easier to read.

File: pkgs/development/libraries/openmpi/default.nix
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix:130]: > configure: error: "yes" or "no" are not valid arguments for --with-ofi-libdir

on aarch64-linux this seems to yield "yes" or "no" !?

Comment (doronbehar) [reply pkgs/development/libraries/openmpi/default.nix]: > > configure: error: "yes" or "no" are not valid arguments for --with-ofi-libdir
> 
> on aarch64-linux this seems to yield "yes" or "no" !?

No, but the configure script parses `--without-ofi-libdir` in an equivalent way to `--with-ofi-libdir=no` which apparently is not supported.. I switched to using `lib.optionals` only for this option.

**Comment by doronbehar** [comment]
> On request: please put the reformatting in its own commit (i.e. put "and more" in a separate commit). That makes the changes easier to read.

Working on it now.

**Comment by doronbehar** [comment]
> > On request: please put the reformatting in its own commit (i.e. put "and more" in a separate commit). That makes the changes easier to read.
> 
> Working on it now.

Done.

**Comment by doronbehar** [comment]
@ofborg build cp2k python3Packages.meep

**Comment by doronbehar** [comment]
CI was green for aarch64-linux before this last force push. I noticed a small mistake in the commit log that is fixed now.

**Comment by doronbehar** [comment]
I'd suggest to handle https://github.com/NixOS/nixpkgs/pull/327444 before merging this.

File: pkgs/by-name/pr/prrte/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/pr/prrte/package.nix:38]:
```suggestion
    patchShebangs ./autogen.pl ./config
```

File: pkgs/development/libraries/openmpi/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/openmpi/default.nix:46]:
```suggestion
    url = "https://www.open-mpi.org/software/ompi/v${lib.versions.majorMinor finalAttrs.version}/downloads/openmpi-${finalAttrs.version}.tar.bz2";
```

**Review by SuperSandro2000 — COMMENTED** [review]
can't say much about most other changes

**Comment by markuskowa** [comment]
mpi4py requires a test to be disabled: https://github.com/NixOS/nixpkgs/pull/275796/commits/c2e58ed8de03d1146267299445ccd5c9d8fe0a6e
btw: mpi4py is the reason the openmpi update was blocked for so long. Otherwise, we could have merged the other PR 6 months ago.

File: pkgs/by-name/pr/prrte/package.nix
Comment (doronbehar) [reply pkgs/by-name/pr/prrte/package.nix]: > nit

I agree.

File: pkgs/development/libraries/openmpi/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/openmpi/default.nix]: Thanks.

**Comment by doronbehar** [comment]
> mpi4py requires a test to be disabled: [c2e58ed](https://github.com/NixOS/nixpkgs/commit/c2e58ed8de03d1146267299445ccd5c9d8fe0a6e) btw: mpi4py is the reason the openmpi update was blocked for so long. Otherwise, we could have merged the other PR 6 months ago.

@markuskowa thanks for explaining. Indeed mpi4py was is in a problematic state. Let's deal with that in https://github.com/NixOS/nixpkgs/pull/327444 .

**Comment by doronbehar** [comment]
@markuskowa I rebased this on top of latest master (with https://github.com/NixOS/nixpkgs/pull/327444 merged) and tested that mpi4py builds fine, and I also checked `python311Packages.h5py-mpi` (which runs tests that use `python312Packages.mpi4py` and `python312Packages.pytest-mpi`), and it built fine as well. Waiting for your review.

**Comment by markuskowa** [comment]
@doronbehar  I'll take a look at it next week. Sorry for the delay.

**Comment by doronbehar** [comment]
Fixed a merge conflict after 63d05d989e73 .

> @doronbehar I'll take a look at it next week. Sorry for the delay.

Still waiting :)

**Review by markuskowa — APPROVED** [review]
I checked and it does look good now. Let's get this merged.

Only one note: I am not a big fan of the "Handle all compiler wrappers consistently & declaratively". It is more complicated than the old shell code version, which is easier to read.

**Comment by doronbehar** [comment]
> I checked and it does look good now. Let's get this merged.
> 
> Only one note: I am not a big fan of the "Handle all compiler wrappers consistently & declaratively". It is more complicated than the old shell code version, which is easier to read.

Hmmm that is the highlight of this PR haha :). I agree it is less readable, but I think it is much harder to maintain a manually written list of 7x3 commands that look almost the same. Also it was annoying for me when I tried to debug this part in the old way, that I had to fully build the package every time I wanted to slightly change something there, due to small typos etc.. No wonder so many files weren't handled in that phase before this commit. With the new code, there should be no typos.

You can also now debug the resulted hook with:

```sh
nix eval --json .\#openmpi.postInstall | jq --raw-output .
```

I'll also note before I'll merge this, that I tried to run a `nixpkgs-review` a few days ago, and most packages built fine, besides the tensorflow related packages which my machine just didn't manage to finish building, even after a few days.

cc me for any breakages :).

File: pkgs/development/libraries/openmpi/default.nix
Comment (FRidh) [inline pkgs/development/libraries/openmpi/default.nix:224]: This comment gets interpolated and therefore requires `pmix` which is not available on Darwin.

Comment (doronbehar) [reply pkgs/development/libraries/openmpi/default.nix:224]: > This comment gets interpolated and therefore requires `pmix` which is not available on Darwin.

pmix is available on Darwin since https://github.com/NixOS/nixpkgs/pull/333016 .

File: pkgs/development/libraries/openmpi/default.nix
Comment (Artturin) [inline pkgs/development/libraries/openmpi/default.nix:178]: What's the reason for using `targetPackages.gfortran` here? It'll only work on non-cross where `targetPackages.gfortran` is the same as `gfortran`? Asking for https://github.com/NixOS/nixpkgs/pull/341269

If you were just following the `targetPackages.stdenv` above then `targetPackages.stdenv` is the same offset as `gfortran`

```
nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.cc
«derivation /nix/store/cy0iv1sk4c60df358j3ff8xb2x0db1qh-aarch64-unknown-linux-gnu-gcc-wrapper-13.3.0.drv»

nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc
«derivation /nix/store/bh6d4yc7dyx1694gz3bx65l0flv1blc2-gcc-wrapper-13.3.0.drv»


nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc.stdenv.hostPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc.stdenv.targetPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.hostPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.targetPlatform.system
"aarch64-linux"
```

Comment (doronbehar) [reply pkgs/development/libraries/openmpi/default.nix:178]: > EDIT: disregard the below, it should be for when `openmpi` goes in `nativeBuildInputs` right?

Correct. I'm not sure how to test that with `nix-repl` like tests as you tried to demonstrate, but yes.

**Pull request merged**
